File: PP/lista3/lista3.ml

let rec print_list = function
  | [] -> ()
  | h::t -> print_int h; print_string " "; print_list t;
;;


let rec print_list_t = function
  | [] -> ()
  | (a, b)::t -> print_string "(";print_int a; print_string ", ";print_int b; print_string ") "; print_list_t t;
;;


let curry f x y z = f(x, y, z);;
(*
  * ('a * 'b * 'c -> 'd) -> 'a -> 'b -> 'c -> 'd
*)

let uncurry f (x, y, z) = f x y z;;
(*
  * ('a -> 'b -> 'c -> 'd) -> 'a * 'b * 'c -> 'd
*)

let rec sumProd l =
  List.fold_left (fun (a1, a2) x -> (a1+x, a2*x)) (0, 1) l
;;


let insertionSort orr l =
  let rec ins n e =
    match n with
      | [] -> [e]
      | h::t -> if orr e h then e::n else h::ins t e
    in
      List.fold_left (fun acc ee -> ins acc ee) [] l
;;


let rec mergeSort orr l =
    let len = List.length l / 2 in
    if len = 0 then l else
      let rec merge ll rr =
        match (ll, rr) with
          | (h1::t1, h2::t2) -> if orr h1 h2 then h1::merge t1 rr else h2::merge ll t2
          | (a, []) -> a
          | ([], a) -> a
      and split l r a =
          if a = 0 then (List.rev l,r) else split (List.hd r :: l) (List.tl r) (a-1)
      in
        let (ll, rr) = split [] l len in merge (mergeSort orr ll) (mergeSort orr rr)

;;


let test = [(2, 199); (1, 2); (1, 123); (3, 25); (2, 99); (3, 0); (4,8886); (1, 322);];;


let testOrder f s =
  fst f <= fst s
;;

 print_list_t (mergeSort testOrder test);;

File: PP/lista3/lista3.scala
object lista3 {
  def printList(args: List[_]): Unit = {
    args.foreach(println)
  }

  def sumProd (l: List[Int]): (Int, Int) = {
    (l.foldLeft ((0, 1))) ((t, x) => (t._1 + x, t._2 * x))
  }

  def insertionSort[A](orr: (A, A) => Boolean, l: List[A]) = {
    def ins (n: List[A], e: A): List[A] = n match {
        case Nil => e::Nil
        case h::t => if (orr (e, h)) e::n else h::(ins (t, e))
    }
    (l foldLeft List[A]()) ((acc, ee) => ins (acc, ee))
  }

  def mergeSort[A](orr: (A, A) => Boolean, l: List[A]):List[A]  = {
      var len = l.length / 2
      if (len == 0) l else {
        def merge (ll: List[A], rr: List[A]): List[A] = (ll, rr) match {
            case (h1::t1, h2::t2) => if (orr(h1, h2)) h1::merge(t1, rr) else h2::merge(ll, t2)
            case (a, Nil) => a
            case (Nil, a) => a
        }

        def split (l: List[A], r: List[A], a: Int): (List[A], List[A]) =
            if (a == 0) (l.reverse, r) else split((r.head :: l), (r.tail), (a-1))

        var (ll, rr) = split(Nil, l, len)
        merge (mergeSort (orr, ll), mergeSort (orr, rr))
      }
  }




  def main(args: Array[String]): Unit = {
    var list = List(List(5,6), List(1,2,3))
    var list2 = List(5, 6, 5, 6, 7, 3, 1, 2)
    var list3 = List(1,2,3,2,1)
    var list4 = List(1,2,3)

    println()
    printList(insertionSort(((a: Int, b: Int) => a <= b ) ,list2))
    printList(mergeSort(((a: Int, b: Int) => a <= b ) ,list2))


  }
}
